src/c/urweb.c

```diff
@@ -57,6 +58,50 @@ int uw_really_write(int fd, const void *buf, size_t len) {
   return 0;
 }
 
+int uw_accept(int sockfd, struct sockaddr *addr, socklen_t *addrlen) {
+  static pthread_mutex_t m = PTHREAD_MUTEX_INITIALIZER;
+
+  int ret;
+  int flags;
+
+  flags = fcntl(sockfd, F_GETFL, 0);
+  if(flags < 0)
+    return flags;
+  ret = fcntl(sockfd, F_SETFL, flags | O_NONBLOCK);
+  if(ret < 0)
+    return ret;
+
+  assert(sockfd < FD_SETSIZE);
+  fd_set sockfds;
+  FD_ZERO(&sockfds);
+  FD_SET(sockfd, &sockfds);
+  ret = select(sockfd+1, &sockfds, NULL, NULL, NULL /* wait infinitely */);
+  if(ret < 0)
+    goto unblock;
+  if(ret == 0) {
+    ret = -1;
+    goto unblock;
+  }
+
+  pthread_mutex_lock(&m);
+  int new_fd = accept(sockfd, addr, addrlen);
+  if(new_fd > 0) {
+    ret = fcntl(new_fd, F_SETFD, FD_CLOEXEC);
+  }
+  else {
+    ret = -1;
+  }
```

Comment:
How about making it so that all the `if`s in this function are for error conditions? You could say

```c
if (new_fd < 0) {
  ret = -1;
  goto unblock;
}
if (fcntl(new_fd, F_SETFD, FD_CLOEXEC) < 0)
  goto unblock;
```